src/utils/typenames.ts

```diff
@@ -7,15 +7,15 @@ import {
 } from 'graphql';
 
 const getTypeNameFromField = (obj: object) =>
-  Object.values(obj).reduce(
-    (all, val) =>
-      typeof val !== 'object' || val === null
-        ? all
-        : val.__typename !== undefined
-        ? { ...all, [val.__typename]: true }
-        : { ...all, ...getTypeNameFromField(val) },
-    {}
-  );
+  Object.values(obj).reduce((all, val) => {
+    if (typeof val !== 'object' || val === null) {
+      return all;
+    }
+
+    return val.__typename !== undefined
+      ? { ...all, [val.__typename]: true, ...getTypeNameFromField(val) }
+      : { ...all, ...getTypeNameFromField(val) };
+  }, {});
```

Comment:
There are some problems here:

- Arrays are not taken into account
- Nested arrays are not taking into account
- We have an object to keep track of types but then also `filter` afterwards

I'd go for a recursive function that isn't exported, like `getTypeNameFromField` right now, but make it completely mutable and filter afterwards, i.e.:

```
interface EntityLike {
  __typename: ?string,
  [key: string]: EntityLike | EntityLike[] | string | number | null
}

const collectTypes = (obj: EntityLike, types: string[]) => {
  if (Array.isArray(obj)) {
    obj.forEach(inner => collectTypes(inner, types));
  } else {
    for (const key in obj) {
      if (object.hasOwnProperty(key)) {
        const val = obj[key];
        if (key === '__typename' && typeof val === 'string') {
          types.push(val);
        } else if (typeof val === 'object') {
          collectTypes(val, types);
        }
      }
    }
  }

  return types;
};

export const gankTypeNamesFromResponse = (response: object) => {
  const typeNames = collectTypes((response as EntityLike), []);
  return typeNames.filter((v, i, a) => a.indexOf(v) === i);
};
```